native/src/main/java/io/ballerina/stdlib/http/api/HttpDispatcher.java

```diff
@@ -390,9 +392,18 @@ private static BObject createRequest(HttpCarbonMessage httpCarbonMessage, BObjec
 
     static BObject getCaller(Resource resource, HttpCarbonMessage httpCarbonMessage,
                              BMap<BString, Object> endpointConfig) {
-        final BObject httpCaller = httpCarbonMessage.getProperty(HttpConstants.CALLER) == null ?
-                ValueCreatorUtils.createCallerObject(httpCarbonMessage) :
-                (BObject) httpCarbonMessage.getProperty(HttpConstants.CALLER);
+        Object availableCallerObj = httpCarbonMessage.getProperty(HttpConstants.CALLER);
+        String resourceAccessor = resource.getBalResource().getAccessor().toUpperCase(Locale.getDefault());
+        if (Objects.nonNull(availableCallerObj)) {
+            BObject caller = (BObject) availableCallerObj;
+            Object currentResourceAccessor = caller.get(HttpConstants.RESOURCE_ACCESSOR);
+            if (Objects.isNull(currentResourceAccessor) ||
+                    HttpUtil.isDefaultResource(((BString) currentResourceAccessor).getValue())) {
+                caller.set(HttpConstants.RESOURCE_ACCESSOR, StringUtils.fromString(resourceAccessor));
+            }
```

Comment:
AFAIR we need to populate the service associated with the current resource to the caller object, to get some configs when we respond. So can we call `HttpUtil.enrichHttpCallerWithConnectionInfo` before returning the caller?

```suggestion
            }
            HttpUtil.enrichHttpCallerWithConnectionInfo(caller, httpCarbonMessage, resource, endpointConfig);
```

Reply:
Fixed with [bf17cfc22072fb7203cfd45ea5be98c0f51f4146](https://github.com/ballerina-platform/module-ballerina-http/pull/1101/commits/bf17cfc22072fb7203cfd45ea5be98c0f51f4146)